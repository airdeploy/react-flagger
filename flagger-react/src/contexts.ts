import {IEntity, IFlaggerConfiguration} from 'flagger/types/Types'
import React from 'react'

export interface IFlaggerCtx {
  entity?: IEntity
  config?: IFlaggerConfiguration
  loading: boolean
  getVariation: (codename: string, entity?: IEntity) => string
  getFlagDetails: (
    codename: string,
    entity?: IEntity
  ) => {variation: string; enabled: boolean; isSampled: boolean; payload: any}
}

export const flaggerCtx = React.createContext<IFlaggerCtx>({
  loading: false,
  getVariation: () => 'off',
  getFlagDetails: () => ({
    variation: 'off',
    enabled: false,
    isSampled: false,
    payload: null,
  }),
})
