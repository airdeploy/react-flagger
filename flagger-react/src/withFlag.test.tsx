import {Logger} from 'flagger'
import React from 'react'
import renderer from 'react-test-renderer'

import {flaggerCtx} from './contexts'
import {withFlag} from './index'

function TestComponent(props: any) {
  return <div>{JSON.stringify(props, null, 2)}</div>
}

describe('withFlag() tests', () => {
  beforeAll(async () => {
    jest.mock('flagger')
  })

  afterAll(async () => {
    jest.clearAllMocks()
  })

  test('logs a warning message and renders wrapped component with empty flags prop if no flags provided', async () => {
    const warnSpy = jest.spyOn(Logger.prototype, 'warn')
    const getFlagDetails = jest.fn()
    const getVariation = jest.fn()
    const entity = {id: '1'}

    const WrappedComponent = withFlag(TestComponent)

    const component = renderer.create(
      <flaggerCtx.Provider
        value={{
          loading: false,
          getVariation,
          getFlagDetails,
        }}>
        <WrappedComponent entity={entity} />
      </flaggerCtx.Provider>
    )
    const expectedRender = renderer.create(
      <TestComponent loading={false} flags={{}} entity={entity} />
    )

    const json = component.toJSON()
    expect(json).toMatchSnapshot()
    expect(json).toEqual(expectedRender.toJSON())
    expect(warnSpy).toBeCalledTimes(1)
    expect(warnSpy.mock.calls[0]).toEqual([
      'withFlag did not receive a valid flag name',
    ])
  })

  test('calls getFlagsDetails() from flaggerCtx', () => {
    const getFlagDetails = jest.fn()
    const getVariation = jest.fn()
    const entity = {id: '1'}

    const WrappedComponent = withFlag(TestComponent, 'bitcoin-pay')

    const component = renderer.create(
      <flaggerCtx.Provider
        value={{
          loading: false,
          getVariation,
          getFlagDetails,
        }}>
        <WrappedComponent entity={entity} />
      </flaggerCtx.Provider>
    )

    expect(component.toJSON()).toMatchSnapshot()
  })

  test('uses default entity if not provided', () => {
    const getFlagDetails = jest.fn()
    const getVariation = jest.fn()
    const entity = {id: '1'}
    const WrappedComponent = withFlag(TestComponent, 'bitcoin-pay')

    const component = renderer.create(
      <flaggerCtx.Provider
        value={{
          entity,
          loading: false,
          getVariation,
          getFlagDetails,
        }}>
        <WrappedComponent />
      </flaggerCtx.Provider>
    )

    expect(component.toJSON()).toMatchSnapshot()
    expect(getFlagDetails.mock.calls[0]).toEqual(['bitcoin-pay', entity])
  })

  test('prop entity overrides default one', () => {
    const getFlagDetails = jest.fn()
    const getVariation = jest.fn()
    const entity = {id: '1'}
    const entity2 = {id: '2'}
    const WrappedComponent = withFlag(TestComponent, 'bitcoin-pay')

    const component = renderer.create(
      <flaggerCtx.Provider
        value={{
          entity,
          loading: false,
          getVariation,
          getFlagDetails,
        }}>
        <WrappedComponent entity={entity2} />
      </flaggerCtx.Provider>
    )

    expect(component.toJSON()).toMatchSnapshot()
    expect(getFlagDetails.mock.calls[0]).toEqual(['bitcoin-pay', entity2])
  })
})
