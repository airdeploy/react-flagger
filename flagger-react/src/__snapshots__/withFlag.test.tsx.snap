// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`withFlag() tests calls getFlagsDetails() from flaggerCtx 1`] = `
<div>
  {
  "loading": false,
  "flags": {},
  "entity": {
    "id": "1"
  }
}
</div>
`;

exports[`withFlag() tests logs a warning message and renders wrapped component with empty flags prop if no flags provided 1`] = `
<div>
  {
  "loading": false,
  "flags": {},
  "entity": {
    "id": "1"
  }
}
</div>
`;

exports[`withFlag() tests prop entity overrides default one 1`] = `
<div>
  {
  "loading": false,
  "flags": {},
  "entity": {
    "id": "2"
  }
}
</div>
`;

exports[`withFlag() tests uses default entity if not provided 1`] = `
<div>
  {
  "loading": false,
  "flags": {}
}
</div>
`;
